src/net/net.zig

```diff
@@ -73,7 +72,140 @@ pub const SocketAddr = union(enum(u8)) {
             .port = addr_port,
         } };
     }
+    pub fn parse_v6(buf: []const u8) !Self {
+        if (buf[0] != '[') {
+            return error.Incomplete;
+        }
+        var ip_port: u16 = 0;
+        var result = Self{
+            .V6 = SocketAddrV6{
+                .scope_id = 0,
+                .port = std.mem.nativeToBig(u16, ip_port),
+                .flowinfo = 0,
+                .ip = Ipv6Addr{ .octets = undefined },
+            },
+        };
+
+        // the new buffer stores just the ip part of the ipv6 without the port
+        var ip_address_buf = buf[1..];
+        var ip_sep_index: usize = 0;
+        while (ip_sep_index < buf.len) : (ip_sep_index += 1) {
+            if (buf[ip_sep_index] == ']') {
+                // loop until it finds the index where the ip ends and port starts
+                ip_address_buf = buf[1..ip_sep_index];
+                ip_port = try std.fmt.parseInt(u16, buf[ip_sep_index + 2 ..], 10);
+                result.V6.port = ip_port;
+                break;
+            } else if (ip_sep_index == buf.len - 1) {
+                return error.Incomplete;
+            }
+        }
+        var ip_slice: *[16]u8 = result.V6.ip.octets[0..];
+
+        var tail: [16]u8 = undefined;
+        var x: u16 = 0;
+        var saw_any_digits = false;
+        var index: u8 = 0;
+        var scope_id = false;
+        var abbrv = false;
+        for (ip_address_buf, 0..) |c, i| {
+            if (scope_id) {
+                if (c >= '0' and c <= '9') {
+                    const digit = c - '0';
+                    {
+                        const ov = @mulWithOverflow(result.V6.scope_id, 10);
+                        if (ov[1] != 0) return error.Overflow;
+                        result.V6.scope_id = ov[0];
+                    }
+                    {
+                        const ov = @addWithOverflow(result.V6.scope_id, digit);
+                        if (ov[1] != 0) return error.Overflow;
+                        result.V6.scope_id = ov[0];
+                    }
+                } else {
+                    return error.InvalidCharacter;
+                }
+            } else if (c == ':') {
```

Comment:
similar to scope id comment - since this relates to parsing characters (ie, saw_any_digits), can you move this block closer to the character parsing logic